core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -123,68 +126,52 @@ package object internal {
   }
 
   // Adapted from https://github.com/typelevel/cats-effect/issues/199#issuecomment-401273282
-  @deprecated(
-    "Replaced by cats.effect.Async.fromFuture. You will need a ContextShift[F].",
-    "0.21.4")
+  @deprecated("Replaced by cats.effect.Async.fromFuture", "0.21.4")
   private[http4s] def fromFuture[F[_], A](f: F[Future[A]])(implicit F: Async[F]): F[A] =
-    f.flatMap { future =>
-      future.value match {
-        case Some(value) =>
-          F.fromTry(value)
-        case None =>
-          F.async { cb =>
-            future.onComplete {
-              case Success(a) => cb(Right(a))
-              case Failure(t) => cb(Left(t))
-            }(direct)
-          }
-      }
-    }
+    F.fromFuture(f)
 
   // Adapted from https://github.com/typelevel/cats-effect/issues/160#issue-306054982
   @deprecated("Use `fromCompletionStage`", since = "0.21.3")
   private[http4s] def fromCompletableFuture[F[_], A](fcf: F[CompletableFuture[A]])(implicit
-      F: Concurrent[F]): F[A] =
+      F: Async[F]): F[A] =
     fcf.flatMap { cf =>
-      F.cancelable { cb =>
-        cf.handle[Unit]((result, err) =>
+      F.async { cb =>
+        F.delay(cf.handle[Unit]((result, err) =>
           err match {
             case null => cb(Right(result))
             case _: CancellationException => ()
             case ex: CompletionException if ex.getCause ne null => cb(Left(ex.getCause))
             case ex => cb(Left(ex))
-          })
-        F.delay { cf.cancel(true); () }
+          })) >>
+          F.delay(Some(F.delay(cf.cancel(true)).void))
       }
     }
 
   private[http4s] def fromCompletionStage[F[_], CF[x] <: CompletionStage[x], A](
       fcs: F[CF[A]])(implicit
       // Concurrent is intentional, see https://github.com/http4s/http4s/pull/3255#discussion_r395719880
-      F: Concurrent[F],
-      CS: ContextShift[F]): F[A] =
+      F: Async[F]): F[A] =
     fcs.flatMap { cs =>
       F.async[A] { cb =>
-        cs.handle[Unit] { (result, err) =>
+        F.delay(cs.handle[Unit] { (result, err) =>
           err match {
             case null => cb(Right(result))
             case _: CancellationException => ()
             case ex: CompletionException if ex.getCause ne null => cb(Left(ex.getCause))
             case ex => cb(Left(ex))
           }
-        }
-        ()
-      }.guarantee(CS.shift)
+        }).as(none[F[Unit]])
+      }
     }
 
   private[http4s] def unsafeToCompletionStage[F[_], A](
       fa: F[A]
-  )(implicit F: Effect[F]): CompletionStage[A] = {
+  )(implicit dispatcher: Dispatcher[F], F: Sync[F]): CompletionStage[A] = {
     val cf = new CompletableFuture[A]()
-    F.runAsync(fa) {
-      case Right(a) => IO { cf.complete(a); () }
-      case Left(e) => IO { cf.completeExceptionally(e); () }
-    }.unsafeRunSync()
+    dispatcher.unsafeRunSync(fa.attemptTap {
```

Comment:
use `unsafeRunAsync`. I need to think about threading the `Dispatcher`

Reply:
Can't see `unsafeRunAsync` on dispatcher, so I did `unsafeToFuture`; would that work the same?

Reply:
`unsafeToFuture` will work everywhere you've got an `unsafeRunSync` but don't need backpressure. We could write an `unsafeRunAsync` that performs the `attemptTap` for you, I think that's going to be a common pattern

Reply:
I agree, were it not for removals I would've had to do that multiple times just in this PR alone.